src/main/java/com/bookstore/service/category/CategoryServiceImpl.java

```diff
@@ -0,0 +1,75 @@
+package com.bookstore.service.category;
+
+import com.bookstore.dto.book.BookResponseDtoWithoutCategoryIds;
+import com.bookstore.dto.category.CategoryRequestDto;
+import com.bookstore.dto.category.CategoryResponseDto;
+import com.bookstore.exception.EntityNotFoundException;
+import com.bookstore.exception.InvalidUpdateDataException;
+import com.bookstore.mapper.CategoryMapper;
+import com.bookstore.model.Category;
+import com.bookstore.repository.CategoryRepository;
+import com.bookstore.service.book.BookService;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+    private final BookService bookService;
+
+    @Override
+    public List<CategoryResponseDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryResponseDto getById(Long id) {
+        return categoryMapper.toDto(categoryRepository.getReferenceById(id));
+    }
+
+    @Override
+    public CategoryResponseDto save(CategoryRequestDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, CategoryRequestDto categoryDto) {
+        Category category = categoryRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("Can't update because there is no "
+                        + "category in the DB with id: " + id));
+        if (categoryDto == null || !isUpdateDataPresent(categoryDto)) {
```

Comment:
Please move all possible fields validations to dtos to not clutter your business code with it